packages/zen-bull-vault/src/EmergencyWithdraw.sol

```diff
@@ -0,0 +1,253 @@
+        zenBullTotalSupplyForEulerWithdrawal = IERC20(_zenBull).totalSupply();
+
+        IERC20(_wPowerPerp).approve(_zenBull, type(uint256).max);
+        IERC20(_usdc).approve(_zenBull, type(uint256).max);
+    }
+
+    /**
+     * @dev receive ETH
+     */
+    receive() external payable {
+        require(msg.sender == weth, "Can't receive ETH from this sender");
+    }
+
+    /**
+     * @notice withdraw ETH deposited into crab
+     * @dev this will give the sender ZBEPR token as ownership for the ETH deposited in Euler pool
+     * @param _zenBullAmount ZenBull amount to redeem
+     * @param _maxEthForWPowerPerp max ETH to pay for flashswapped oSQTH amount
+     */
+    function emergencyWithdrawEthFromCrab(uint256 _zenBullAmount, uint256 _maxEthForWPowerPerp)
+        external
+    {
+        uint256 crabToRedeem = _zenBullAmount.wdiv(zenBullTotalSupplyForCrabWithdrawal).wmul(
+            IZenBullStrategy(zenBull).getCrabBalance()
+        );
+        (, uint256 wPowerPerpInCrab) = IZenBullStrategy(zenBull).getCrabVaultDetails();
+        uint256 wPowerPerpToRedeem =
+            crabToRedeem.wmul(wPowerPerpInCrab).wdiv(IERC20(crab).totalSupply());
+
+        zenBullTotalSupplyForCrabWithdrawal =
+            zenBullTotalSupplyForCrabWithdrawal.sub(_zenBullAmount);
+
+        IERC20(zenBull).transferFrom(msg.sender, address(this), _zenBullAmount);
+        _mint(msg.sender, _zenBullAmount);
+
+        _exactOutFlashSwap(
+            weth,
+            wPowerPerp,
+            3000,
+            wPowerPerpToRedeem,
+            _maxEthForWPowerPerp,
+            uint8(FLASH_SOURCE.EMERGENCY_WITHDRAW_ETH_FROM_CRAB),
+            abi.encodePacked(crabToRedeem, wPowerPerpToRedeem)
+        );
+
+        uint256 payout = IERC20(weth).balanceOf(address(this));
+        IWETH9(weth).withdraw(payout);
+        payable(msg.sender).sendValue(payout);
+
+        emit Withdraw(
+            msg.sender, _zenBullAmount, crabToRedeem, wPowerPerpToRedeem, payout, _zenBullAmount
+        );
+    }
+
+    /**
+     * @notice repay a portion of Euler debt and withdraw WETH collateral to this contract based on ZenBullEulerRecovery amount
+     * @dev this will revert of WETH to withdraw from Euler is greater than max, or limit price breach the tolerance %
+     * @dev if all ZenBullEulerRecovery are burnt, the ETH withdrawl will be activated
+     * @param _wethToWithdraw amount of WETH to withdraw from Euler
+     * @param _maxEthForUsdc max ETH to pay for 1 USDC
+     * @param _poolFee ETH/USDC Uni v3 pool fee
+     */
+    function emergencyRepayEulerDebt(
+        uint256 _wethToWithdraw,
+        uint256 _maxEthForUsdc,
+        uint24 _poolFee
+    ) external {
+        require(
+            _wethToWithdraw <= MAX_WETH_PER_DEBT_REPAY,
+            "WETH to withdraw is greater than max per repay"
+        );
+
+        uint256 ratio = _wethToWithdraw.wdiv(IEulerEToken(eToken).balanceOfUnderlying(zenBull));
```

Comment:
might be impossible to make this ratio 100%, because the eToken balanceOfUnderlying may change constantly (even just in the time from off chain calc to onchain execution)

I know I said fine to do either via explicit amounts, but maybe its better to have the input be a percentage, so can start with 10%, then 20%, then 50%, then 100% and explicitly try to withdraw 100%.

Reply:
And you can calculate the amounts in the same way, as usdcToRepay (and for wethToWithdraw), using the explicit ratio.